test/Test/Util/FS.hs

```diff
@@ -228,6 +237,74 @@ approximateEqStream (UnsafeStream infoXs xs) (UnsafeStream infoYs ys) =
       (Finite, Finite)     ->  xs == ys
       (_, _)               -> False
 
+{-------------------------------------------------------------------------------
+  List directory
+-------------------------------------------------------------------------------}
+
+data DirEntry a = Directory a | File a
+  deriving stock (Show, Eq, Ord, Functor)
+
+-- | List all files and directories in the given directory and recusively in all
+-- sub-directories.
+listDirectoryRecursive ::
+     Monad m
+  => HasFS m h
+  -> FsPath
+  -> m (Set (DirEntry FsPath))
+listDirectoryRecursive hfs = go Set.empty (mkFsPath [])
+  where
+    go !acc relPath absPath = do
+        pcs <- listDirectory hfs absPath
+        foldlM (\acc' -> go' acc' relPath absPath) acc pcs
+
+    go' !acc relPath absPath pc = do
+        let
+          p = mkFsPath [pc]
+          relPath' = relPath </> p
+          absPath' = absPath </> p
+        isFile <- doesFileExist hfs absPath'
+        if isFile then
+          pure (File relPath' `Set.insert` acc)
+        else do
+          isDirectory <- doesDirectoryExist hfs absPath'
+          if isDirectory then
+            go (Directory relPath' `Set.insert` acc) relPath' absPath'
+          else
+            error "impossible"
```

Comment:
Under what conditions would this happen? I suppose if the testing directory contains a node that isn't a file or directory, which can happen, but indicates a failure... so should perhaps this error have a slightly clearer message, in case it does actually happen?